allennlp/modules/elmo.py

```diff
@@ -670,3 +689,156 @@ def create_cached_cnn_embeddings(self, tokens: List[str]) -> None:
+        hidden_size = W.shape[1]
+        fc_layer = torch.nn.Linear(hidden_size, vocab_size)
+
+        fc_layer.weight.data.copy_(
+            torch.FloatTensor(W))
+        fc_layer.bias.data.copy_(
+            torch.FloatTensor(b))
+
+        return fc_layer
+
+    def _chunked_log_probs(self,
+                           activation: torch.Tensor,
+                           word_targets: torch.Tensor) -> torch.Tensor:
+        """
+        Do the softmax in chunks so the gpu ram doesn't explode.
+
+        Parameters
+        ----------
+        activation : ``torch.Tensor``, required
+            Results of the ELMo embedder.
+        word_targets : ``torch.Tensor``, required
+            The word_ids returned from `batch_to_ids`.
+
+        Returns
+        -------
+        ``torch.Tensor`` with all of the log probabilities.
+        """
+        all_logprobs = []
+        num_chunks = (activation.size(0) - 1) // self.chunk_size + 1
+        for activation_chunk, target_chunk in zip(
+                torch.chunk(activation, num_chunks, dim=0),
+                torch.chunk(word_targets, num_chunks, dim=0)):
+
+            assert activation_chunk.size()[:2] == target_chunk.size()[:2]
+            targets_flat = target_chunk.view(-1)
+            time_indexer = torch.arange(
+                0, targets_flat.size(0),
+                out=target_chunk.data.new(
+                    targets_flat.size(0))) % target_chunk.size(1)
+            batch_indexer = torch.arange(
+                0, targets_flat.size(0),
+                out=target_chunk.data.new(
+                    targets_flat.size(0))) / target_chunk.size(1)
+            all_logprobs.append(
+                F.log_softmax(self.fc_layer(activation_chunk), 2)[
+                batch_indexer, time_indexer, targets_flat].view(
+                    *target_chunk.size()))
+
+        return torch.cat(all_logprobs, 0)
+
+    def forward(self,
+                bilm_inputs: Dict[str, Union[torch.Tensor, List[torch.Tensor]]],
+                word_inputs: torch.Tensor,
+                aggregation_fun: str = None) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Passes the ELMo embeddings to a softmax layer and calculates
+        the log probabilities.
+
+        Parameters
+        ----------
+        bilm_inputs : ``Dict[str, Union[torch.Tensor, List[torch.Tensor]]]``, required
+            The output from `_ElmoBiLm`.
+        word_inputs : ``torch.Tensor``, required
+            The output from `batch_to_ids` when `
+        aggregation_fun : ``str``, optional
+            An aggregation function that aggregates the forward and backward
+            tensors. Default is `None` where no aggregation is done.
+
+        Returns
+        -------
+        A tuple of `log_probs` and `mask_without_bos_eos`.
+        """
+
+        activations = bilm_inputs['activations']
+        mask = bilm_inputs['mask']
+
+        activations = activations[-1]  # (batch, BOS + L + EOS, [fwd, bwd])
```

Comment:
Please use full variable names and don't re-use variable names which have different types. e.g

```python
# shape: (batch_size, sentence_length + 2, 2 * encoder_dim)
final_activations = activations[-1]
```